core/store/benches/finalize_bench.rs

```diff
@@ -0,0 +1,271 @@
+//! Benchmarks for the in-memory preparation of writes before the commit, aka
+//! `finalize`.
+//!
+//! All writes are batched to a DB transaction before being committed. This
+//! involves copying around the data and deserializing it.
+//!
+//! Since this is done on the hot path, whereas the actual commit can often be
+//! done in a different thread, we have this benchmark to test the speed.
+//! 
+//! Right now, the module only contains test for `ShardChunk` and
+//! `PartialEncodedChunk`. These have been observed to become quite large and
+//! have ab impact on the overall client performance.
+
+#[macro_use]
+extern crate bencher;
+
+use bencher::{black_box, Bencher};
+use borsh::BorshSerialize;
+use near_chain::Chain;
+use near_chunks::ShardsManager;
+use near_crypto::{InMemorySigner, KeyType, Signer};
+use near_primitives::hash::CryptoHash;
+use near_primitives::merkle::{merklize, MerklePathItem};
+use near_primitives::receipt::{ActionReceipt, DataReceipt, Receipt, ReceiptEnum};
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::sharding::{
+    ChunkHash, EncodedShardChunk, PartialEncodedChunk, PartialEncodedChunkPart,
+    PartialEncodedChunkV2, ReceiptProof, ReedSolomonWrapper, ShardChunk, ShardChunkHeader,
+    ShardChunkHeaderV3, ShardChunkV2, ShardProof,
+};
+use near_primitives::transaction::{Action, FunctionCallAction, SignedTransaction};
+use near_primitives::types::AccountId;
+use near_primitives::validator_signer::InMemoryValidatorSigner;
+use near_store::DBCol;
+use rand::prelude::SliceRandom;
+
+/// `ShardChunk` -> `StoreUpdate::insert_ser`.
+///
+/// ~10ms or faster for ~24MB receipts
+fn benchmark_write_shard_chunk(bench: &mut Bencher) {
+    let transactions = vec![];
+    let receipts = create_benchmark_receipts();
+    let chunk_hash: ChunkHash = CryptoHash::default().into();
+    let chunk = create_shard_chunk(&chunk_hash, transactions, receipts);
+
+    let chunks = spread_in_memory(chunk);
+
+    let store = near_store::test_utils::create_test_store();
+    bench.iter(|| {
+        let mut update = store.store_update();
+        update
+            .insert_ser(DBCol::Chunks, chunk_hash.as_ref(), &chunks.choose(&mut rand::thread_rng()))
+            .unwrap();
```

Comment:
I thought you'd also call finalize here but it seems there isn't one on store update. Is this line the core piece of finalize implementation?

Reply:
Ah yeah it's a bit confusing.

The`finalize` I'm targeting is this: https://github.com/near/nearcore/blob/21686cfb7f0393648aa0edbcaee2c2e7d960a6b3/chain/chain/src/store.rs#L2943

It's pretty long and does many different operations on store_update. Ideally, I would call that function for the benchmark. But it requires an entire `ChainStore` and way too many details to be right for the finalization to succeed. (Note how in my benchmark I get away with fake hashes in most places.)

This commit only covers a fraction of what's happening in there. Specifically, it targets `insert_ser` because that's where (1) I found a bottleneck and (2) I found an easy optimization. So a benchmark for this specific operation is the most useful right now.
Furthermore, for the benchmarks I just picked two (known-to-be-slow) examples of data we insert in finalize.

I will maybe add more benchmarks in this file which will also cover other operations, such as `increment_rc` or `set`. Depends on whether I discover them to also be a bottleneck for `finalize` or not.